functionary/core/api/v1/views/task.py

```diff
@@ -67,16 +80,26 @@ def get_serializer_class(self):
         },
         parameters=HEADER_PARAMETERS,
     )
-    def create(self, request, *args, **kwargs):
-        request_serializer = self.get_serializer(data=request.data)
+    def create(self, request: Request, *args, **kwargs):
+        data = request.data
+
+        # Remove list elements that wrap singular values when multipart content
+        data = request.data.dict()
+
+        request_serializer: Union[
+            TaskCreateByIdSerializer, TaskCreateByNameSerializer
+        ] = self.get_serializer(data=data)
+
         request_serializer.is_valid(raise_exception=True)
 
         request_serializer.save(
-            creator=self.request.user,
-            environment=self.get_environment(),
+            creator=self.request.user, environment=self.get_environment()
         )
 
+        _handle_file_parameters(request, request_serializer)
```

Comment:
There's a race condition here because the task is already saved when you get to this point, so if this file upload doesn't happen fast enough the tasking will have run already without the file being there.

I think there's a couple things may want to do or thing about doing here.

1) We should probably separate the task publication (the thing that's actually going to result in the tasking reaching the runner) from the task instance creation.  Right now the publish happens automatically in the `post_create`.  If we instead make it an explicit step, where you have to call `task.publish()`, then you can control when that happens and make sure you only call it after the file parameters have been handled.

2) Given the above, we'll probably need some error handling for after you save the task but before you publish in the event that the file upload fails.  Not just in terms of the API response error, but marking the task itself as having errored.  This is kind of not ideal though because it's not really the tasking that failed (the function didn't run and fail) it was submitting the task that failed.  It seems like the ideal behavior would be to have never actually created the task if the file upload is unsuccessful, but we need some info from the task (sort-of) in order to do the file handling.  The easiest solution to that is to wrap the task creation and file upload stuff in a transaction.  There are some risks there given that the file upload could take a long-ish time in some circumstances, but it's probably the thing to do for now.

Maybe do the transaction thing now and then if you want to put a `#TODO` comment in to come back and address the race condition because it has implications in other parts of the code, I would be okay with that.

Reply:
I had a belated realization that I should check how we're doing this in the UI view side of things, which revealed what is probably the right / easier solution to all of this for now....

You could probably just not call `request_serializer.save()` and instead just instantiate a `Task` using the `validated_data`.  Then call `task.save()` after you've done the file handling.  That's basically what happens on the UI side.

Reply:
Went with the third recommendation. I think one follow on task from this issue could be to centralize the `Task` creation logic. I've spread the logic between both the API and the UI view, and they can be combined or maybe moved to a helper method for the Task model